internal/handlers/sjson/sjson_test.go

```diff
@@ -152,125 +171,199 @@ func testJSON(t *testing.T, testCases []testCase, newFunc func() sjsontype) {
-					docs = append(docs, doc)
-				}
+			if doc := b.ReturnFieldsSelector; doc != nil {
+				docs = append(docs, doc)
+			}
 
-				if doc := b.ReturnFieldsSelector; doc != nil {
-					docs = append(docs, doc)
-				}
+		case *wire.OpReply:
+			docs = append(docs, b.Documents...)
+		}
 
-			case *wire.OpReply:
-				docs = append(docs, b.Documents...)
+		for _, doc := range docs {
+			if doc.ValidateData() != nil {
+				continue
 			}
 
-			for _, doc := range docs {
-				if doc.ValidateData() != nil {
-					continue
-				}
+			args := make([]any, 0, 2)
 
+			if needDocument {
 				j, err := MarshalSingleValue(doc)
 				require.NoError(f, err)
+				args = append(args, j)
+			}
 
+			if needSchema {
 				sch, err := marshalSchemaForDoc(doc)
 				require.NoError(f, err)
-
-				f.Add(string(j), string(sch))
-				n++
+				args = append(args, sch)
 			}
+
+			f.Add(args...)
+			n++
+		}
+	}
+
+	return n
+}
+
+func fuzzJSONWithFixedDocuments(f *testing.F, testCases []testCase, newFunc func() sjsontype) {
+	type documentTest struct {
+		value sjsontype
+		j     string
+	}
+	documents := make([]documentTest, 0, len(testCases))
+
+	for _, tc := range testCases {
+		if tc.jErr != "" || tc.v == nil {
+			continue
 		}
+		sch := must.NotFail(json.Marshal(tc.sch))
+		f.Add(string(sch))
+		documents = append(documents, documentTest{
+			value: tc.v,
+			j:     tc.j,
+		})
+	}
 
+	if _, ok := newFunc().(*documentType); ok && !testing.Short() {
+		needDocument := false
+		needSchema := true
+		n := addRecordedFuzzDocs(f, needDocument, needSchema)
 		f.Logf("%d recorded documents were added to the seed corpus", n)
 	}
 
-	f.Fuzz(func(t *testing.T, j, jsch string) {
+	f.Fuzz(func(t *testing.T, jsch string) {
 		t.Parallel()
 
-		// raw "null" should never reach UnmarshalJSON due to the way encoding/json works
-		if j == "null" {
-			t.Skip()
-		}
-
-		// no reason to fuzz all the possible input for schemas,
-		// so if sch is not a valid schema, skip the test
 		var sch elem
 		if err := json.Unmarshal([]byte(jsch), &sch); err != nil {
-			t.Skip()
+			t.Skip("invalid schema")
 		}
 
-		// j may not be a canonical form.
-		// We can't compare it with MarshalJSON() result directly.
-		// Instead, we compare with round-trip result.
-
-		v := newFunc()
-		tc := testCase{
-			sch: &sch,
-			j:   j,
-		}
-		if err := unmarshalJSON(v, &tc); err != nil {
-			t.Skip()
-		}
-
-		// Temporary hack, should be removed once we improve our validation.
-		// TODO https://github.com/FerretDB/FerretDB/issues/1273
-		{
-			d, ok := fromSJSON(v).(*types.Document)
-			if !ok {
-				t.Skip()
+		for _, d := range documents {
+			actualV := newFunc()
+			tc := testCase{
+				sch: &sch,
+				j:   d.j,
 			}
-			if err := d.ValidateData(); err != nil {
-				t.Skip()
+			if err := unmarshalJSON(actualV, &tc); err == nil {
+				// if there was no error (which is a rare occasion),
+				// check that the documents match each other
+				assertEqualForFuzz(t, d.value, actualV)
```

Comment:
What was your observation on how rare it was? :thinking: 

From CI, I see that we have pretty good number running against fuzz target against `FuzzArrayWithFixedDocuments` where you made improvement :muscle: 

```
task: [fuzz] go test -run=XXX -fuzz=FuzzArrayWithFixedDocuments    -fuzztime=15s ./internal/handlers/sjson/
fuzz: elapsed: 0s, gathering baseline coverage: 0/3 completed
fuzz: elapsed: 0s, gathering baseline coverage: 3/3 completed, now fuzzing with 1 workers
fuzz: elapsed: 3s, execs: 30008 (10002/sec), new interesting: 18 (total: 21)
fuzz: elapsed: 6s, execs: 61123 (10372/sec), new interesting: 35 (total: 38)
fuzz: elapsed: 9s, execs: 85691 (8187/sec), new interesting: 48 (total: 51)
fuzz: elapsed: 12s, execs: 114037 (9451/sec), new interesting: 56 (total: 59)
fuzz: elapsed: 15s, execs: 145002 (10321/sec), new interesting: 71 (total: 74)
fuzz: elapsed: 16s, execs: 145002 (0/sec), new interesting: 71 (total: 74)
PASS
ok  	github.com/FerretDB/FerretDB/internal/handlers/sjson	15.760s
```

Reply:
It's rare in comparison with generating something that will not pass that check. :sweat_smile:
The fuzz runner needs to generate a schema that will allow the document to be unmarshalled correctly.
I removed the comment since it looks confusing.

Reply:
Got it :+1: